src/Bicep.Core/Emit/ExternalInputFunctionReferenceVisitor.cs

```diff
@@ -0,0 +1,106 @@
+        this.variablesContainingExternalInput = ImmutableHashSet.CreateBuilder<VariableDeclarationSyntax>();
+    }
+
+    public override void VisitVariableAccessSyntax(VariableAccessSyntax syntax)
+    {
+        var symbol = semanticModel.GetSymbolInfo(syntax);
+
+        if (symbol is DeclaredSymbol declaredSymbol && declaredSymbol.Type is ErrorType)
+        {
+            return;
+        }
+
+        switch (symbol)
+        {
+            case ParameterAssignmentSymbol parameterAssignmentSymbol:
+                Visit(parameterAssignmentSymbol.DeclaringSyntax);
+                break;
+
+            case VariableSymbol variableSymbol:
+                Visit(variableSymbol.DeclaringSyntax);
+                break;
+        }
+
+        base.VisitVariableAccessSyntax(syntax);
+    }
+
+
+    public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
+    {
+        if (string.Equals(syntax.Name.IdentifierName, LanguageConstants.ExternalInputBicepFunctionName, LanguageConstants.IdentifierComparison))
+        {
+            this.externalInputReferences.TryAdd(syntax, this.externalInputReferences.Count);
+            if (this.targetParameterAssignment is not null)
+            {
+                this.parametersContainingExternalInput.Add(this.targetParameterAssignment);
+            }
+            
+            if (this.targetVariableDeclaration is not null)
+            {
+                this.variablesContainingExternalInput.Add(this.targetVariableDeclaration);
+            }
+        }
+
+        base.VisitFunctionCallSyntax(syntax);
+    }
+
+    public static ExternalInputReferences CollectExternalInputReferences(SemanticModel model)
+    {
+        var visitor = new ExternalInputFunctionReferenceVisitor(model);
+
+        foreach (var paramAssignment in model.Root.ParameterAssignments)
+        {
+            var declaringSyntax = paramAssignment.DeclaringParameterAssignment;
+            visitor.targetParameterAssignment = declaringSyntax;
+            declaringSyntax.Accept(visitor);
+        }
+
+        foreach (var variableDeclaration in model.Root.VariableDeclarations)
+        {
+            var declaringSyntax = variableDeclaration.DeclaringVariable;
+            visitor.targetVariableDeclaration = declaringSyntax;
+            declaringSyntax.Accept(visitor);
+        }
+
+        return new ExternalInputReferences(
+            ParametersReferences: visitor.parametersContainingExternalInput.ToImmutable(),
+            VariablesReferences: visitor.variablesContainingExternalInput.ToImmutable(),
+            ExternalInputIndexMap: visitor.externalInputReferences.ToImmutable()
+        );
+    }
+}
+
+public record ExternalInputReferences(
+    // parameters that contain external input function calls
+    ImmutableHashSet<ParameterAssignmentSyntax> ParametersReferences,
+    // variables that contain external input function calls
+    ImmutableHashSet<VariableDeclarationSyntax> VariablesReferences,
```

Comment:
Is it possible to use symbols rather than syntax? Not a huge deal if it's a complex refactor.

Reply:
done